### PR TITLE
TS-5022: reduce memory allocation in clientcert lookup

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

**Comment by persiaAziz** [comment]
please review @SolidWallOfCode This will reduce the unnecessary allocations. This checks if the client cert filename and path under txn_cnf has the same value as the global defaults.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1363/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1256/ for details.

**Comment by zwoop** [comment]
This has to go into 7.1.0 as well, since we already landed previous tries :).

**Comment by zwoop** [comment]
@shinrich @SolidWallOfCode Is this landing for 7.1.x?

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -5050,9 +5036,26 @@ HttpSM::do_http_server_open(bool raw)
       opt.set_sni_servername(host, len);
     }
 
-    ats_scoped_str clientCert(
-      (Layout::relative_to(t_state.txn_conf->client_cert_filepath, t_state.txn_conf->client_cert_filename)));
-    opt.set_client_certname(clientCert);
+    SSLConfig::scoped_config params;
+    ats_scoped_str clientCert;
```
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: Why is this declared twice?

**Comment by SolidWallOfCode** [comment]
I think I want a different approach on this. I'll talk with @persiaAziz.

**Comment by persiaAziz** [comment]
I was thinking to have one overridable config for the client cert containing the partial or absolute path to avoid the memory allocation in Layout, but that will break the convention of having separate configs for path and filename. It could be also confusing for the user

**Comment by SolidWallOfCode** [comment]
The problem (as @persiaAziz patiently explained) is the client certificate is overridable which means it cannot be reliable precomputed when `remap.config` is parsed. In theory a plugin could load a distinct client certificate every single remap. Unfortunately this means checking every time. Persia said she'd look at having a single path instead of two pieces to simplify the code. I'll do more review once that's ready.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1427/ for details.

**Comment by persiaAziz** [comment]
@SolidWallOfCode, @shinrich   Please review. 
In summary,
* proxy.config.ssl.client.cert.path has been completely removed.
* proxy.config.ssl.client.cert.filename has been renamed to proxy.config.ssl.client.cert.file. 
* proxy.config.ssl.client.cert.file is overridable
* The user needs to specify the fullpath in proxy.config.ssl.client.cert.file.
* It does not check if the given path is absolute or not because if the path is not absolute then concatenating (default)path and filename is again going to end up in allocating memory on each remap.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1320/ for details.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1428/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1321/ for details.

**Comment by jpeach** [comment]
Whilst the configuration of SSL keys and certificates is a bit unfortunate, I think that it is important to be consistent.

Separately, you have to define the base directory for relative paths, otherwise the current working directory of whatever does the loading becomes an ABI.

**Comment by zwoop** [comment]
@SolidWallOfCode Is this going to go in for 7.1.x ?

**Comment by zwoop** [comment]
@SolidWallOfCode Ping? Also, seems this needs to be rebased now.

**Comment by persiaAziz** [comment]
Not needed

**Pull request closed without merging**